File: src/main/kotlin/ua/com/lavi/usergenerator/factory/UserFactory.kt
package ua.com.lavi.usergenerator.factory

import ua.com.lavi.usergenerator.FakeUser
import ua.com.lavi.usergenerator.utils.PasswordGenerator
import ua.com.lavi.usergenerator.utils.Transliterator
import java.io.BufferedReader
import java.util.*

/**
 * Created by Oleksandr Loushkin
 */

abstract class UserFactory {

    private val transliterator: Transliterator = Transliterator()

    private val passwordGenerator: PasswordGenerator = PasswordGenerator()

    abstract fun getLastname(): String

    abstract fun getName(): String

    abstract fun getPatronomyc(): String

    fun getUser(): FakeUser {
        val lastname = getLastname()
        val name = getName()
        val patronomyc = getPatronomyc()

        val fullname = "$lastname $name $patronomyc"
        val username = transliterator.transliterate(fullname.replace(" ", "_").toLowerCase())
        val password = passwordGenerator.generatePassword()
        val fakeUser = FakeUser(
                lastname,
                name,
                patronomyc,
                fullname,
                username,
                password)

        return fakeUser
    }


    fun getUsers(count: Int): List<FakeUser> {
        val fakeUsers: MutableList<FakeUser> = ArrayList(count)
        var x = 0
        while (x < count) {
            fakeUsers.add(getUser())
            x++
        }
        return fakeUsers
    }

    protected fun readResourceLines(filename: String): List<String> {
        return this.javaClass.classLoader.getResourceAsStream(filename).bufferedReader().
                use(BufferedReader::readLines)
    }

    protected fun randomStringFromList(list: List<String>): String {
        val rnd = Random().nextInt(list.size)
        return list[rnd]
    }


}

File: src/main/scala/ua/com/lavi/usergenerator/factory/UserFactory.scala
package ua.com.lavi.usergenerator.factory

import ua.com.lavi.usergenerator.FakeUser
import ua.com.lavi.usergenerator.utils.{PasswordGenerator, Transliterator}

import scala.collection.mutable.ArrayBuffer
import scala.io.Source
import scala.util.Random

/**
  * Created by Oleksandr Loushkin
  */

abstract class UserFactory {

  val transliterator: Transliterator = new Transliterator()

  private val passwordGenerator: PasswordGenerator = new PasswordGenerator()

  def getLastname(): String

  def getName(): String

  def getPatronomyc(): String

  def getUser(): FakeUser = {
    val lastname: String = getLastname()
    val name: String = getName()
    val patronomyc: String = getPatronomyc()

    val fullname = s"$lastname $name $patronomyc"
    val username = transliterator.transliterate(fullname.replace(" ", "_").toLowerCase())
    val password = passwordGenerator.generatePassword()
    val fakeUser = FakeUser(
      lastname,
      name,
      patronomyc,
      fullname,
      username,
      password)

    fakeUser
  }

  protected def readResourceLines(filename: String): Seq[String] = {
    Source.fromResource(filename).getLines().toSeq
  }

  def randomStringFromList(list: Seq[String]): String = {
    list(new Random().nextInt(list.size))
  }

  def getUsers(count: Int): Seq[FakeUser] = {
    val fakeUsers:ArrayBuffer[FakeUser] = ArrayBuffer()
    for (_ <- 0 until count) {
      fakeUsers.append(getUser())
    }

    fakeUsers
  }

}
